ion-schema/src/model/constraints/all_of.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::internal_traits::{
+    LoaderContext, ValidateInternal, ValidationContext, WriteAsIsl, WriteContext,
+};
+use crate::ion_schema_version::Versioned;
+use crate::model::constraints::{ConstraintName, ReadConstraint};
+use crate::model::type_argument::TypeArgument;
+use crate::model::{TypeDefinitionBuilder, VersionedTypeArgumentList};
+use crate::result::IonSchemaResult;
+use crate::{IonSchemaElement, IslVersion, ViolationRecorder};
+use ion_rs::{Element, ValueWriter};
+use std::ops::ControlFlow;
+
+impl ConstraintName for AllOf {
+    const CONSTRAINT_NAME: &'static str = "all_of";
+}
+
+/// Represents the `all_of` constraint (ref. [ISL 1.0], [ISL 2.0]).
+///
+/// [ISL 1.0]: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#all_of
+/// [ISL 2.0]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#all_of
+#[derive(Debug, PartialEq, Clone)]
+pub struct AllOf {
+    type_arguments: Vec<TypeArgument>,
+}
+
+impl AllOf {
+    fn new(mut type_arguments: Vec<TypeArgument>) -> Self {
+        // Sorting the vec allows us to get Bag-like equality semantics.
+        // TODO: Replace the vec with a bag or set, or come up with a less hacky way to sort the type arguments.
+        type_arguments.sort_by_cached_key(|item| format!("{item:?}"));
```

Comment:
Since you're already sorting the `Vec`, you could use something like [`slice::partition_dedup`](https://doc.rust-lang.org/std/primitive.slice.html#method.partition_dedup) to get from Bag to Set.

Reply:
Using a set would be possible, but it would be a compromise rather than the desired state.